antlr-kotlin-target/src/main/resources/org/antlr/v4/tool/templates/codegen/Kotlin/Kotlin.stg

```diff
@@ -633,26 +633,53 @@ cases(tokens) ::= <<
 <tokens:{t | <t.name>}; separator=", "> -> >>
 
 InvokeRule(r, argExprsChunks) ::= <<
-this.state = <r.stateNumber>
-<if(r.labels)><r.labels:{l | <labelref(l)> = }><endif><r.escapedName>(<if(r.ast.options.p)><r.ast.options.p><if(argExprsChunks)>,<endif><endif><argExprsChunks>)
+state = <r.stateNumber>
+<if(r.labels)>
+
+if (true) {
+    val _ctx = <r.escapedName>(<if(r.ast.options.p)><r.ast.options.p><if(argExprsChunks)>, <endif><endif><argExprsChunks>)
+    <r.labels:{l | <labelref(l)> = _ctx}; separator="\n">
+}
+<else>
+<r.escapedName>(<if(r.ast.options.p)><r.ast.options.p><if(argExprsChunks)>, <endif><endif><argExprsChunks>)
+<endif>
 >>
 
 MatchToken(m) ::= <<
-this.state = <m.stateNumber>
-<if(m.labels)><m.labels:{l | <labelref(l)> = }><endif>match(<m.name>)
+state = <m.stateNumber>
+<if(m.labels)>
+
+if (true) {
+    val _token = match(<m.name>)
+    <m.labels:{l | <labelref(l)> = _token}; separator="\n">
+}
+<else>
+match(<m.name>)
+<endif>
 >>
 
 MatchSet(m, expr, capture) ::= "<CommonSetStuff(m, expr, capture, false)>"
 
 MatchNotSet(m, expr, capture) ::= "<CommonSetStuff(m, expr, capture, true)>"
 
 CommonSetStuff(m, expr, capture, invert) ::= <<
-this.state = <m.stateNumber>
-<if(m.labels)><m.labels:{l | <labelref(l)> = }>_input.LT(1)<endif>
+state = <m.stateNumber>
+<if(m.labels)>
+
+if (true) {
```

Comment:
It would be a duplicated comment, but the reasoning is evident once you understand how the template works.

Without the local scoping, the `_tokens` variable would be redefined and thus it would produce an error.

Reply:
Understood